tests/activity.go

```diff
@@ -54,6 +55,139 @@ import (
+
+	var out string
+	err = workflowRun.Get(ctx, &out)
+	s.Equal(2, actionCalled)
+	s.NoError(err)
+}
+
+func (s *ActivityTestSuite) TestActivityScheduleToClose_FiredDuringActivityRun() {
+	// We have activity that always fails.
+	// We have backoff timers and schedule_to_close activity timeout happens while activity is running.
+	// activity will be scheduled twice.
+	// "schedule_to_close" timer should fire while activity is running for a second time
+	initialRetryInterval := time.Second * 1
+	workingInterval := time.Second * 2
+	scheduleToCloseTimeout := 7 * time.Second
+	startToCloseTimeout := 3 * time.Second
+
+	activityRetryPolicy := &temporal.RetryPolicy{
+		InitialInterval:    initialRetryInterval,
+		BackoffCoefficient: 1,
+		MaximumInterval:    time.Second * 20,
+		MaximumAttempts:    10,
+	}
+	var lastActivityRun time.Time
+	var workflowCompleteRun time.Time
+
+	activityCompleted := 0
+	activityFunction := func() (string, error) {
+		println(fmt.Sprintf("Activity #%d", activityCompleted+1))
+		time.Sleep(workingInterval)                        //nolint:forbidigo
+		activityErr := errors.New("bad-luck-please-retry") //nolint:goerr113
+		lastActivityRun = time.Now().UTC()
+		activityCompleted += 1
+		return "", activityErr
+	}
+
+	workflowFn := func(ctx workflow.Context) (string, error) {
+		var ret string
+		err := workflow.ExecuteActivity(workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
+			DisableEagerExecution:  true,
+			StartToCloseTimeout:    startToCloseTimeout,
+			ScheduleToCloseTimeout: scheduleToCloseTimeout,
+			RetryPolicy:            activityRetryPolicy,
+		}), activityFunction).Get(ctx, &ret)
+		var activityError *temporal.ActivityError
+		s.True(errors.As(err, &activityError))
+		s.Equal(enumspb.RETRY_STATE_NON_RETRYABLE_FAILURE, activityError.RetryState())
+		var timeoutError *temporal.TimeoutError
+		s.True(errors.As(activityError.Unwrap(), &timeoutError))
+		s.Equal(enumspb.TIMEOUT_TYPE_SCHEDULE_TO_CLOSE, timeoutError.TimeoutType())
+
+		workflowCompleteRun = time.Now().UTC()
+		// schedule to close timeout should fire while last activity is still running.
+		s.Equal(2, activityCompleted)
+		return "done!", nil
+	}
+
+	s.worker.RegisterWorkflow(workflowFn)
+	s.worker.RegisterActivity(activityFunction)
+
+	workflowOptions := sdkclient.StartWorkflowOptions{
+		ID:                 "functional-test-schedule_to_close_failed_while_running",
+		TaskQueue:          s.taskQueue,
+		WorkflowRunTimeout: 10 * time.Second,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	workflowRun, err := s.sdkClient.ExecuteWorkflow(ctx, workflowOptions, workflowFn)
+	s.NoError(err)
+
+	var out string
+	err = workflowRun.Get(ctx, &out)
+	workflowCompleteRun = time.Now().UTC()
+	// we expect activities to be executed 3 times
+	s.Eventually(func() bool { return activityCompleted == 3 }, time.Second*10, time.Millisecond*500)
+	s.True(lastActivityRun.After(workflowCompleteRun))
```

Comment:
You're using second timer granularity that will reduce the chance of flakes but this isn't perfect. In certain scenarios, these tests will be flaky. I recommend relying less on timing and more on sending messages between channels if possible. It's tricky to get right because there are timers that are run internally by the server and you can't control that from tests.

Reply:
`sending messages between channels if possible`
in this case I don't see how it is possible. I not just need to wait when activity is completed. I need activity to be completed _after_ proter shedule_to_close timeout fired, but before it was fired in the previous version of the code.